### DISCUSSION TITLE
Started using Distributed data - get ddatareplicator dead letter messages

### DISCUSSION THREAD

**Discussion by giovanni-ferrari** [akkadotnet/akka.net#7344] [Q&A] [answered]
I started using Distributed Data to store temporary information that should be resilient across the cluster (i'm using distributed data in conjunction with sharding). 

Since then I started seeing some [INFO] logs from akka.net, that repeats continuosly.

Sep 27 09:36:07 ubuntu service.service[3831814]: [INFO][09/27/2024 07:36:07.329Z][Thread 0017][akka://Service/user/ddataReplicator] Message [Status] from [akka.tcp://Service@xxx.xxx.xxx.xx0:9997/user/ddataReplicator#43346742] to [akka://Service/user/ddataReplicator] was not delivered. [9] dead letters encountered. If this is not an expected behavior then [akka://Service/user/ddataReplicator] may have terminated unexpectedly. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'. Message content: Status(chunk=0, totalChunks=1, digest={Data1->Google.Protobuf.ByteString,Data2->Google.Protobuf.ByteString,Data3->Google.Protobuf.ByteString,})

Any idea why this is happening, and if it can be related to any configuration issue?

**Comment by Aaronontheweb** [comment]
Is the replicator not running on your current node? Because that's why you're getting the dead letter - the actor at your local path `akka://Service/user/ddataReplicator` doesn't exist. If the replicators are syncing based on role types, the issue might be that the replicator on `xxx.xxx.xxx.xx0:9997` expects there to be one running locally on your other node as well, but there isn't one. If you didn't specify a role setting for the replicator, then that would also cause this issue - no role means _everyone_ gets all the the replication messages.

**Comment by giovanni-ferrari** [comment, marked as answer]
Thank you Aaron, I confirm the problem was related to roles configuration

**Reply by Aaronontheweb** [reply]
Glad to hear you got to the bottom of it

**Reply by giovanni-ferrari** [reply]
Hi Aaron,
After the configuration was fixed, I did not receive the "dead letters" error message for a long time, but it has recently started appearing again.
What happens now is that the application runs correctly for about 10 hours, then the dead letter messages start showing up again (only on one node in a cluster of three). Even if the node is restarted, the issue persists.

I also tried setting 'RecreateOnFailure' to true, but it had no effect.

Message [Status] from [akka.tcp://service@service1:9997/user/DataReplicator#973567920] to [akka://Service/user/DataReplicator] was not delivered. [168] dead letters encountered, of which 157 were not logged. The counter will be reset now. If this is not an expected behavior then [akka://Service/user/DataReplicator] may have terminated unexpectedly. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'. Message content: Status(chunk=0, totalChunks=1, digest={Missions->Google.Protobuf.ByteString,Charging->Google.Protobuf.ByteString,Alarms->Google.Protobuf.ByteString,}),

**Reply by Aaronontheweb** [reply]
@giovanni-ferrari was there a recent upgrade or change inside your system that could have prompted this?

**Reply by giovanni-ferrari** [reply]
It happened on a new installation - but the code that initializes the data replicator and use it to store/retrieve data has not changed. Neither the roles configuration for the cluster nodes.

**Reply by Aaronontheweb** [reply]
Got it - would you mind sharing what one of your DData writes / reads looks like? It could be innocuous: write needs a majority + 1 quorum and gets it, thus it completes, but other nodes who received the write request are still sending back ACKs that don't need to be processed any longer.

**Reply by giovanni-ferrari** [reply]
Hi Aaron, 
this is how we read and write ddata

```csharp
        var cluster = Akka.Cluster.Cluster.Get(Context.System);
        var replicator = DistributedData.Get(Context.System).Replicator;
        var savepointLWWRegisterKey = new LWWRegisterKey<long>(_configuration.Id);
        var response = await replicator.Ask<IGetResponse>(Dsl.Get(savepointLWWRegisterKey, ReadLocal.Instance));
        T lastReadSavepoint = defaultValue;
        bool foundSavepoint = false;
        LWWRegister<T>? lastReadLWWRegisterSavePoint = null;
        if (response.IsFound)
        {
            try
            {
                lastReadLWWRegisterSavePoint = response.Get(savepointLWWRegisterKey);
                lastReadSavepoint = lastReadLWWRegisterSavePoint.Value;
                foundSavepoint = true;
            }
            catch (KeyNotFoundException ex)
            {
                _logger.LogError(ex, "Found no key matching id");
            }
        }
// some code that retrieves a new "lastReadSavepoint"

        lastReadLWWRegisterSavePoint = lastReadLWWRegisterSavePoint?.WithValue(cluster.SelfUniqueAddress, lastReadSavepoint) ?? new LWWRegister<long>(cluster.SelfUniqueAddress, lastReadSavepoint);
        var updateResponse = await replicator.Ask<IUpdateResponse>(Dsl.Update(savepointLWWRegisterKey, lastReadLWWRegisterSavePoint, WriteLocal.Instance, _ => lastReadLWWRegisterSavePoint));
        updateResponse.ThrowOnFailure();
```

**Reply by Aaronontheweb** [reply]
@giovanni-ferrari ok, so you're writing with a `WriteLocal` consistency which is the lowest kind - so the `DeadLetter`s are not happening as a result of a "high write consistency with `ReadWriteAggregator`s shutting down if it completes early" theory.

Looking at your `DeadLetter` logs:

```text
Message [Status] from [akka.tcp://service@service1:9997/user/DataReplicator#973567920] to [akka://Service/user/DataReplicator] was not delivered. 
```

So on the local node where this `DeadLetter` was logged, there was no actor found at the location `/user/DataReplicator` - can you confirm that:

1. The replicator was indeed started at that location - you can use a [`pbm actor hierarchy` command](https://cmd.petabridge.com/articles/commands/actor-commands.html) to confirm this, but generally you should probably be able to grok it from the source too.
2. That the replicator was not killed for some reason - if it was killed as a result of a supervision directive you would have seen a stream of `LogLevel.Error` messages spewing out of it first most likely.
